acme/tests/messages_test.py

```diff
@@ -287,7 +287,7 @@ class UpdateRegistrationTest(unittest.TestCase):
     def test_empty(self):
         from acme.messages import UpdateRegistration
         jstring = '{"resource": "reg"}'
-        self.assertEqual(jstring, UpdateRegistration().json_dumps())
+        self.assertEqual('{}', UpdateRegistration().json_dumps())
```

Comment:
Why did this change?

Reply:
In ACMEv1 drafts -01 and -02, request bodies sent by ACME clients [sometimes needed to contain the `resource` field](https://datatracker.ietf.org/doc/html/draft-ietf-acme-acme-02#section-6.3). For challenges, there was the `type` field.

From what I can glean, Certbot was still doing this in ACMEv2 days until it was eventually fixed in #7176.

What I think is going on with this test, is that it was originally written according for the ACMEv1 -01 -02 drafts: the `resource` field is expected when `UpdateRegistration` is serialized to JSON.

Since in this PR, ACMEv1 support is gone, the serialization of an empty `UpdateRegistration` is now an empty JSON object.

Now that I have written out this novel, I realize that maybe we should look into whether we can just eliminate `ResourceMixin`, `TypeMixin` and `type` and `resource` fields in the relevant classes. If so, we should include them in the list of new deprecations in 1.30.0.

Reply:
I do want to say, I'm open to breaking that out into a separate PR (should be straightforward since it was added in a separate PR), or just building onto this branch, whichever you prefer.

Reply:
I am having some difficulty ripping out the classes and attributes mentioned earlier. I think the problem is that I don't understand what `typ` and `resource` actually do in the `acme` library and how it uses `josepy`. 

Also unsure whether we need to keep using `jose.TypedJSONObjectWithFields`, which is the class that's inserting the `type` field, which then gets reversed by `TypeMixin`. The purpose of `@Challenge.register` and `@Directory.register` is lost on me as well.

I wonder if somebody else wants to have a crack at cleaning these last bits up, or providing guidance on how to move forward here?

Reply:
I'm happy to take a crack at it, but in that case, I'll approve this now so we can get it in and look at that separately, unless you think it needs to be coupled to everything else here for some reason.

Reply:
(as an aside, looks like we are once again running into https://github.com/certbot/certbot/issues/1297)

Reply:
To further document some things I've learned (hopefully understood correctly):

- I think we can remove `resource` and `ResourceMixin` cleanly because it's only used as an ACMEv1 field and `josepy` doesn't otherwise know it exists
- We rely on `TypedJSONObjectWithFields` for `acme.challenges.Challenge` to register and then to look up challenge types later. If we got rid of `type` and `TypeMixin`, we'd have to replace the registration and lookup of challenges with a different API.

Questions:

- Not clear why `ChallengeResponse` uses `TypedJSONObjectWithFields` or whether  `@ChallengeResponse.register` functionally does anything 
-`Directory.register` doesn't actually use `TypedJSONObjectWithFields` and wouldn't be affected.  But it's not clear to me what the function does. It writes to `_REGISTERED_TYPES` and then nobody ever reads it? Is it there to prevent duplicate registrations?